File: code/programs/go/unix-tools/rm_tool.go
// =========================================================================
// rm — Remove Files or Directories
// =========================================================================
//
// The `rm` utility removes files and directories. It is one of the most
// dangerous Unix commands — there is no "undo" or "trash" for rm.
//
// # Basic usage
//
//   rm file.txt                 Remove a single file
//   rm -f file.txt              Force: don't error if file doesn't exist
//   rm -r dir/                  Recursively remove directory and contents
//   rm -rf dir/                 Force recursive removal (common pattern)
//   rm -v file.txt              Verbose: print what's being removed
//
// # The -r flag (recursive)
//
// Without -r, rm refuses to remove directories:
//
//   rm mydir/    => ERROR: mydir is a directory
//
// With -r, rm walks the directory tree depth-first, removing files and
// subdirectories before removing the parent:
//
//   rm -r mydir/
//     remove mydir/file1.txt
//     remove mydir/sub/file2.txt
//     remove mydir/sub/
//     remove mydir/
//
// # Architecture
//
//   rm.json (spec)              rm_tool.go (this file)
//   ┌──────────────────┐      ┌────────────────────────────────┐
//   │ flags: -f,-r,-d  │      │ for each file arg:             │
//   │ -i,-I,-v         │─────>│   check type (file/dir)        │
//   │ variadic FILE    │      │   remove with appropriate fn   │
//   └──────────────────┘      │   print if verbose             │
//                             └────────────────────────────────┘

package main

import (
	"fmt"
	"io"
	"os"

	clibuilder "github.com/adhithyan15/coding-adventures/code/packages/go/cli-builder"
)

// =========================================================================
// runRm — the testable core of the rm tool
// =========================================================================

func runRm(specPath string, argv []string, stdout io.Writer, stderr io.Writer) int {
	// Step 1: Create the parser from the spec.
	parser, err := clibuilder.NewParser(specPath, argv)
	if err != nil {
		fmt.Fprintf(stderr, "rm: %s\n", err)
		return 1
	}

	// Step 2: Parse the arguments.
	result, err := parser.Parse()
	if err != nil {
		fmt.Fprintf(stderr, "%s\n", err)
		return 1
	}

	// Step 3: Handle the result.
	switch r := result.(type) {

	case *clibuilder.HelpResult:
		fmt.Fprintln(stdout, r.Text)
		return 0

	case *clibuilder.VersionResult:
		fmt.Fprintln(stdout, r.Version)
		return 0

	case *clibuilder.ParseResult:
		// Extract flags.
		force := getBool(r.Flags, "force")
		recursive := getBool(r.Flags, "recursive")
		removeDir := getBool(r.Flags, "dir")
		verbose := getBool(r.Flags, "verbose")

		// Extract file paths.
		files := getStringSlice(r.Arguments, "files")

		exitCode := 0
		for _, file := range files {
			// Check if the file/dir exists.
			info, err := os.Lstat(file)
			if err != nil {
				if force {
					// With -f, silently ignore nonexistent files.
					continue
				}
				fmt.Fprintf(stderr, "rm: cannot remove '%s': %s\n", file, err)
				exitCode = 1
				continue
			}

			// Handle directories.
			if info.IsDir() {
				if recursive {
					// os.RemoveAll removes the directory and everything inside it.
					// This is the equivalent of `rm -r`.
					err = os.RemoveAll(file)
				} else if removeDir {
					// os.Remove only works on empty directories.
					// This is the equivalent of `rm -d` (same as rmdir).
					err = os.Remove(file)
				} else {
					fmt.Fprintf(stderr, "rm: cannot remove '%s': Is a directory\n", file)
					exitCode = 1
					continue
				}
			} else {
				// Regular file, symlink, etc. — just remove it.
				err = os.Remove(file)
			}

			if err != nil {
				if !force {
					fmt.Fprintf(stderr, "rm: cannot remove '%s': %s\n", file, err)
					exitCode = 1
				}
				continue
			}

			if verbose {
				fmt.Fprintf(stdout, "removed '%s'\n", file)
			}
		}

		return exitCode

	default:
		fmt.Fprintf(stderr, "rm: unexpected result type: %T\n", result)
		return 1
	}
}

File: code/programs/rust/unix-tools/src/rm_tool.rs
//! # rm — Remove Files or Directories
//!
//! This module implements the business logic for the `rm` command.
//! rm removes files and directories. There is no "undo" — use with care.
//!
//! ## Modes
//!
//! ```text
//!     rm file.txt         Remove a file
//!     rm -r dir/          Recursively remove directory and contents
//!     rm -f missing.txt   Force: ignore nonexistent files
//!     rm -d empty_dir/    Remove empty directory (like rmdir)
//! ```

use std::fs;
use std::path::Path;

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Remove a single file or directory.
///
/// # Parameters
/// - `path`: the path to remove
/// - `recursive`: if true, remove directories recursively
/// - `dir_mode`: if true, allow removing empty directories
/// - `force`: if true, silently ignore nonexistent paths
///
/// # Returns
/// `Ok(())` on success, `Err(message)` on failure.
pub fn remove_path(path: &str, recursive: bool, dir_mode: bool, force: bool) -> Result<(), String> {
    let file_path = Path::new(path);

    // Check if the path exists.
    let metadata = match fs::symlink_metadata(file_path) {
        Ok(m) => m,
        Err(_) => {
            if force {
                return Ok(()); // Silently ignore nonexistent files
            }
            return Err(format!("rm: cannot remove '{}': No such file or directory", path));
        }
    };

    if metadata.is_dir() {
        if recursive {
            // Remove the entire directory tree.
            fs::remove_dir_all(file_path)
                .map_err(|e| format!("rm: cannot remove '{}': {}", path, e))
        } else if dir_mode {
            // Remove only if empty (like rmdir).
            fs::remove_dir(file_path)
                .map_err(|e| format!("rm: cannot remove '{}': {}", path, e))
        } else {
            Err(format!("rm: cannot remove '{}': Is a directory", path))
        }
    } else {
        // Regular file or symlink.
        fs::remove_file(file_path)
            .map_err(|e| format!("rm: cannot remove '{}': {}", path, e))
    }
}

// ---------------------------------------------------------------------------
// Unit tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use std::env;

    fn temp_path(name: &str) -> String {
        let dir = env::temp_dir();
        dir.join(format!("rm_test_{}", name))
            .to_string_lossy()
            .into_owned()
    }

    #[test]
    fn remove_file() {
        let path = temp_path("file");
        fs::write(&path, "data").unwrap();
        assert!(remove_path(&path, false, false, false).is_ok());
        assert!(!Path::new(&path).exists());
    }

    #[test]
    fn remove_nonexistent_without_force() {
        let result = remove_path("/tmp/rm_test_nonexistent_xyz", false, false, false);
        assert!(result.is_err());
    }

    #[test]
    fn remove_nonexistent_with_force() {
        let result = remove_path("/tmp/rm_test_nonexistent_xyz", false, false, true);
        assert!(result.is_ok());
    }

    #[test]
    fn remove_dir_without_recursive() {
        let path = temp_path("dir_no_r");
        let _ = fs::remove_dir_all(&path);
        fs::create_dir_all(&path).unwrap();
        fs::write(format!("{}/file.txt", path), "data").unwrap();
        let result = remove_path(&path, false, false, false);
        assert!(result.is_err());
        fs::remove_dir_all(&path).ok();
    }

    #[test]
    fn remove_dir_recursive() {
        let path = temp_path("dir_r");
        let _ = fs::remove_dir_all(&path);
        fs::create_dir_all(format!("{}/sub", path)).unwrap();
        fs::write(format!("{}/sub/file.txt", path), "data").unwrap();
        assert!(remove_path(&path, true, false, false).is_ok());
        assert!(!Path::new(&path).exists());
    }

    #[test]
    fn remove_empty_dir_with_d() {
        let path = temp_path("empty_d");
        let _ = fs::remove_dir_all(&path);
        fs::create_dir_all(&path).unwrap();
        assert!(remove_path(&path, false, true, false).is_ok());
        assert!(!Path::new(&path).exists());
    }
}
